File: built-in plugins dsp/mb-piano-wurlitzer.cpp
/**
 * MB Wurlitzer
 * Category : instrument
 * Type     : piano
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Vintage Wurlitzer electric piano
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_PIANO_WURLITZER_H
#define MB_PIANO_WURLITZER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbPianoWurlitzer : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-piano-wurlitzer";
    static constexpr const char* PLUGIN_NAME    = "MB Wurlitzer";
    static constexpr const char* PLUGIN_TYPE    = "piano";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float drive = 0.3f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbPianoWurlitzer() = default;
    ~MbPianoWurlitzer() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.drive = std::clamp(params.drive, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Wurlitzer
        return input;
    }
};

#endif // MB_PIANO_WURLITZER_H

File: built-in plugins dsp/mb-piano-wurlitzer.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbPianoWurlitzerPlugin: PluginDefinition = {
    id: 'mb-piano-wurlitzer', slug: 'mb-piano-wurlitzer', name: 'MB Wurlitzer', category: 'instrument', type: 'piano', version: '1.0.0',
    description: 'Vintage Wurlitzer electric piano', author: 'Max Booster', grade: 'A',
    oscillators: [{ type: 'triangle', detune: 0, gain: 0.6 }, { type: 'square', detune: 0, gain: 0.15 }],
    envelope: { attack: 0.002, decay: 0.5, sustain: 0.4, release: 0.5 },
    parameters: [
      { id: 'drive', name: 'Drive', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { drive: 0.3, volume: 0.8 },
  };

export default MbPianoWurlitzerPlugin;
